DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/README.md

```diff
@@ -0,0 +1,286 @@
+
+![Naive System Waveform](./assets/waveform_naive.png)
+
+### Optimized Kernel
+
+To make the memory system more efficient, we can make three changes:
+
+1. Use the right-most array dimension for unrolled accesses, (that is, store the **column** dimension in the right-most array dimension).
+
+2. Choose the number of banks to a power-of-2 using `constexpr`-math functions. This line of code computes the closest power-of-2 above the given number. We use this power-of-2 as the number of banks in our optimized memory system. Since this design performs up to 5 parallel memory accesses, a minimum of 5 banks are required. Since the next power-of-2 is 8, this design should be specified with 8 banks.
+   > ```
+   > constexpr size_t kNumBanks = fpga_tools::Pow2(fpga_tools::CeilLog2(kNumRows));
+   > ```
+
+3. Change the loop induction variables to match with the memory access pattern and result in the correct indices to `array2d` as assigned in change 1.
+
+The `OptimizedKernel` implements these changes, and executes the following memory access pattern:
+
+![Optimized Kernel Memory Access Pattern](./assets/access_pattern_optimized.gif)
+
+Note that in the Kernel Memory Viewer, although the same bank bits ($b_4$, $b_3$, $b_2$) are used, this time, the memory partitioning is exactly what desired. We can deduce a memory layout using the same table:
+
+|              | Address | Bank bits | Bank # |
+|--------------|---------|-----------|--------|
+| array2d[0][0] | 0x0000  | 0b000     | 0      |
+| array2d[0][1] | 0x0004  | 0b001     | 1      |
+| array2d[0][2] | 0x0008  | 0b010     | 2      |
+| array2d[0][3] | 0x000C  | 0b011     | 3      |
+| array2d[0][4] | 0x0010  | 0b100     | 4      |
+| array2d[0][5] | 0x0014  | 0b010     | 5      |
+| array2d[0][6] | 0x0018  | 0b011     | 6      |
+| array2d[0][7] | 0x001C  | 0b100     | 7      |
+| array2d[1][0] | 0x0020  | 0b000     | 0      |
+| array2d[1][1] | 0x0024  | 0b001     | 1      |
+| ...           |         |           |        |
+
+Because that we are only accessing the first 5 columns of `array2d`, that provides sufficient knowledge for the compiler to optimize out the unused memory banks - bank 5 to 7. As a result, we have the following memory layout:
+
+![Optimized Kernel Memory Layout](./assets/memory_system_optimized.png)
+
+Review the memory access pattern above, we can easily find out that within each loop iteration, each of the memory banks is access precisely twice: 1 load and 1 store. Note that though the `idx` loop also performs loads on each bank, since the values to be loaded are already accessed in the `col` loop, these loads are shared. This time, the memory accesses do not surpass the physical limit of the memory block hardware, and thus, we have eliminated the memory bottleneck. In the meantime, observe that the Kernel Memory Viewer shows pipelined never-stall LSUs for our optimized memory system, and the compiler successfully optimized out the unused memory banks.
```

Comment:
> Note that though the `idx` loop also performs loads on each bank, since the values to be loaded are already accessed in the `col` loop, these loads are shared.

this should be mentioned earlier, perhaps even the first paragraph when you talk about the `row` loop and the first `col` loop.